### PR TITLE
Add support for comment syntax to route detector

### PR DESCRIPTION
**Pull request by JamesNK**
String syntax can be set using a comment:

```cs
// language=Route
var s = "/product/{name}";
```

Razor pages uses a comment to enable route syntax. This PR adds support for detecting comments.

### CODE REVIEW

**Review by captainsafia — COMMENTED** [review]
Is it possible to apply this onto a method that takes a route parameter that isn't marked with the embedded language attribute?

For example, is:

```csharp
static MapAction(string routePattern, Action foo);
// language=Route
MapAction("/{id}", () => ...);
```

**Comment by JamesNK** [comment]
No, that isn't supported. `StringSyntaxAttribute` should be used there.

This is only for placing a comment on a local variable. Background for why I'm adding this: a Razor route, registered with @page directive, is set to a local variable. That local variable will have this comment to enable route tooling features.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePatternCompletionProvider.cs [hunk 1]
```diff
@@ -252,19 +256,24 @@ private static void ProvidePolicyNameCompletions(EmbeddedCompletionContext conte
         context.AddIfMissing("float", suffix: null, "Matches a valid float value in the invariant culture.", WellKnownTags.Keyword, parentOpt);
         context.AddIfMissing("guid", suffix: null, "Matches a valid Guid value.", WellKnownTags.Keyword, parentOpt);
         context.AddIfMissing("long", suffix: null, "Matches any 64-bit integer.", WellKnownTags.Keyword, parentOpt);
-        context.AddIfMissing("minlength", suffix: null, "Matches a string with a length greater than, or equal to, the constraint argument.", WellKnownTags.Keyword, parentOpt);
-        context.AddIfMissing("maxlength", suffix: null, "Matches a string with a length less than, or equal to, the constraint argument.", WellKnownTags.Keyword, parentOpt);
-        context.AddIfMissing("length", suffix: null, @"The string length constraint supports one or two constraint arguments.
+
+        // The following constraints are only available for HTTP route matching.
+        if (context.Options == RouteOptions.Http)
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePatternCompletionProvider.cs:261]: This change is part of supporting component routes. I made it in this PR so I'd have a way to test the option is being detected.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePatternAnalyzer.cs [hunk 2]
```diff
@@ -52,7 +52,7 @@ public void Analyze(SemanticModelAnalysisContext context)
             else
             {
                 var token = child.AsToken();
-                if (!RouteStringSyntaxDetector.IsRouteStringSyntaxToken(token, context.SemanticModel, cancellationToken))
+                if (!RouteStringSyntaxDetector.IsRouteStringSyntaxToken(token, context.SemanticModel, cancellationToken, out var options))
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/RoutePatternAnalyzer.cs:55]:
```suggestion
                if (!RouteStringSyntaxDetector.IsRouteStringSyntaxToken(token, context.SemanticModel, cancellationToken, out var _))
```

Discard? Looks like it's not used...

File: src/Framework/AspNetCoreAnalyzers/test/RouteEmbeddedLanguage/RoutePatternAnalyzerTests.cs [hunk 3]
```diff
@@ -4,13 +4,38 @@
...
+    [Fact]
+    public async Task CommentOnString_ReportResults()
+    {
+        var source = TestSource.Read(
+@"
+class Program
+{
+    static void Main()
+    {
+        // language=Route
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/test/RouteEmbeddedLanguage/RoutePatternAnalyzerTests.cs:24]: Do we have a test that checks that a `language=Route,Http` comment will allow processing the length-specifc constraints and the `language=Route, Component` won't?

**Review by captainsafia — APPROVED** [review]

**Pull request merged**